common/ast/src/parser/rule/util.rs

```diff
@@ -0,0 +1,84 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use nom::branch::alt;
+use nom::combinator::map;
+use nom::error::ErrorKind;
+use nom::error::ParseError;
+use nom::IResult;
+
+use crate::parser::ast::Identifier;
+use crate::parser::token::Token;
+use crate::parser::token::TokenKind;
+
+pub type Input<'a> = &'a [Token<'a>];
+
+pub fn satisfy<'a, F, Error>(cond: F) -> impl Fn(Input<'a>) -> IResult<Input<'a>, &'a Token, Error>
+where
+    F: Fn(&Token) -> bool,
+    Error: ParseError<Input<'a>>,
+{
+    move |i| match i.get(0).map(|t| {
```

Comment:
Maybe we need to check length of `i` here?

Reply:
if `i.len() == 0` then the `get(0)` will return `None`